Tmux: add module

stites opened the pull request:
*See PR history for original, WIP comments.*

This commit adds the tmux program to home-manager and wraps Tmux with any `extraConfig` included in nix.

In addition to configuring Tmux, a user may specify tmux plugins from nixpkgs. These can be included in the list of `plugins` and can either be a package (all tmux plugins live under `nixpkgs.tmuxPlugins.*`), or an object which includes the plugin and an `extraConfig` which will be run immediately after sourcing the tmux plugin.

Finally, this commit introduces two nested programs which may be enabled which depend on tmux: tmuxp and tmuxinator. These do not have the ability to be configured, although this may be a future contribution.

--

Unaddressed point (for a future contributor): there is a vim plugin which is required to navigate between vim and tmux tabs. Should these be treated as separate configurations? It seems like home-manager has the ability to, at the very least, fire off a warning if this vim-plugin is not detected in vim and home manager is enabled. At most, I could see this being an option to enable on tmux (`vim-integration = true`).

rycee commented on modules/programs/tmux.nix:
```diff
@@ -0,0 +1,123 @@
+    mkMerge [
+      {
+        home.packages = [ cfg.package ]
+          ++ (if cfg.tmuxinator.enable then [ pkgs.tmuxinator ] else [])
```
There is a convenience function [`optional`](https://github.com/NixOS/nixpkgs/blob/master/lib/lists.nix#L189-L199) that changes this to a more humane
```nix
home.packages =
  [ cfg.package ]
  ++ optional cfg.tmuxinator.enable pkgs.tmuxinator
  ++ optional cfg.tmuxp.enable pkgs.tmuxp;
```
😉

rycee commented on modules/programs/tmux.nix:
```diff
@@ -0,0 +1,123 @@
+      };
+
+      extraConfig = mkOption {
+        type = types.either types.path types.lines;
```
I'd suggest having simply `types.lines` a user wanting to put the extra configuration in a separate file can just use `builtins.readFile` which doesn't add terribly much noise.

rycee commented:
Thanks for being such a nice contributor reading the contribution file and everything! Brings a tear to my eye 😀

> Nested program configurations (tmuxinator and tmuxp) make sense, IMO, since these programs depend on tmux. I don't think it's too necessary -- I figured I'd ask your thoughts.

I know very little about tmux but I imagine, based on their names, that these program only make sense to use together with tmux. In that case it would make sense to put them under the `programs.tmux` namespace.

On the other hand, is it reasonable for a user to want to manage their own tmux installation (i.e. have `programs.tmux.enable == false`) but still want Home Manager to manage their tmuxinator or tmuxp configuration? I imagine this would be a pretty rare occurrence. But in that case they would have, e.g.,
```nix
programs.tmux.enable = false;
programs.tmux.tmuxinator.enable = true;
```
which is not so nice in my opinion.

> There is a vim plugin which is required to navigate between vim and tmux tabs. Should these be treated as separate configurations? It seems like home-manager has the ability to, at the very least, fire off a warning if this vim-plugin is not detected in vim and home manager is enabled. At most, I could see this being an option to enable on tmux (vim-integration = true).

Yeah, you can do either warnings or assertions in a reasonable way in HM. For this, though, I think it is reasonable to just add a Boolean option to the tmux module called `programs.tmux.enableVimIntegration` that, when enabled, will set up the vim plugin in a suitable way. I would suggest not adding warning or errors here because I imagine a reasonable user may want to not have this close integration between vim and tmux and then they shouldn't be forced to see a warning each time they run home-manager 🙂

stites commented:
Thank you for the code review! I am a little tied up at the moment and will get to this later.

stites commented:
is there a way to test in home-manager? If no, should I document my configurations somewhere?

stites commented:
@rycee, I have everything working on my box -- it's very exciting!!!

I'd like to do one more thing and allow users to configure tmux completely in nix. I'm not sure if there are some guidelines to follow, but I've started this off by creating three types (and one helper-function) here:
- [`setOption`](https://github.com/rycee/home-manager/pull/388/files#diff-abc51f5360e9cfc918eba7a95afc239dR57)
- [`bindOption`](https://github.com/rycee/home-manager/pull/388/files#diff-abc51f5360e9cfc918eba7a95afc239dR50)
- [`unbindOption`](https://github.com/rycee/home-manager/pull/388/files#diff-abc51f5360e9cfc918eba7a95afc239dR65)
- helper function [`mkTmuxOption`](https://github.com/rycee/home-manager/pull/388/files#diff-abc51f5360e9cfc918eba7a95afc239dR25)
I'm, once again, stuck and could use some guidance: Is this unnecessary / overkill? Are the types too verbose? How would I create a heterogeneous list of these three types?

Managing plugins and extraConfig are also now done and ready for review! From what I can tell tmuxp and tmuxinator do not take any extra configurations (and I don't really use them anymore, so the next person can tackle that beast).

Thanks so much!
-s

stites commented on modules/programs/tmux.nix:
```diff
@@ -0,0 +1,203 @@
+      sensibleOnTop = mkOption {
+        type = types.bool;
+        default = false;
+        description = "run the sensible plugin at the top of your .conf (to overwrite configs)";
```
this phrasing could actually be a little better.

rycee replied:
Would it be possible to simply always including the sensible plugin? Is it something that everybody uses?

stites replied:
I don't think so -- I wasn't even aware of it when I first got started with tmux plugins. It's _probably_ best to do this, but I really can't give you a good answer. Reflecting back on this, however, I think this should default to `true`

stites commented on modules/programs/tmux.nix:
```diff
@@ -0,0 +1,203 @@
+      plugins = mkOption {
+        type = types.listOf (types.either types.package pluginModule);
+        description = ''
+          List of tmux plugins to include in your .tmux.conf
```
I need to add a note that these are added to the bottom of the `.tmux.conf`.

stites commented on modules/programs/tmux.nix:
```diff
@@ -0,0 +1,203 @@
+        assertions = [
+          (let
+            not = x: if x then false else true;
+            badPlugins = filter (p: not (hasPrefix "tmuxplugin" (pluginName p))) cfg.plugins;
```
This conditional is a little brittle. Tmux plugins are all prefixed this way, but may not be named like this in the future. A better condition would be checking if the package is in the `pkgs.tmuxPlugins.*` namespace, but I wasn't sure how to do this.

rycee replied:
Could use `hasAttr (pluginName p) pkgs.tmuxPlugins`?

kalbasit commented:
@stites @rycee what's blocking this from getting merged?

rycee commented on modules/programs/tmux.nix:
```diff
@@ -0,0 +1,203 @@
+        type = types.string;
+        description = "Type of ${name}.";
+        default = "";
+        example = "\"-key\" becomes \"${name}-key\".";
```
Hmm, I don't know anything about tmux. If I put `bind.specifier = ''"-key" becomes "bind-key"''` in my configuration tmux will do something sensible?

rycee commented on modules/programs/tmux.nix:
```diff
@@ -0,0 +1,203 @@
+    };
+  };
+
+  unbindOption = mkTmuxOption "unbind" { };
```
Some of these don't seem to be used.

stites commented on modules/programs/tmux.nix:
```diff
@@ -0,0 +1,203 @@
+  cfg = config.programs.tmux;
+
+  mkChildPackageType = p: (
+   let
```
oops!

stites commented on modules/programs/tmux.nix:
```diff
@@ -0,0 +1,203 @@
+        ${(concatMapStringsSep "\n\n" (p: ''
+            # ${pluginName p}
+            # ---------------------
+            ${if hasExtraConfig p then p.extraConfig else ""}
```
This doesn't work, I think, because the type is `either (types.either types.package pluginModule)` where a plugin module is a record of package + extra config.

stites commented:
I think I've addressed everything @rycee. I've updated this PR according to contribution guidelines and moved the long description to be the main PR comment.

A short summary of things that did not change:
- `${if hasAttr "extraConfig" p then p.extraConfig else ""}` remains in place of `${p.extraConfig}`. The nix language seems to be erasing the `types.either types.package pluginModule` type which causes some confusion here, but when I pass in a `types.package`, home-manager fails to build.
- the `hasPrefix "tmuxplugin" (pluginName p)` is staying over `hasAttr (pluginName p) pkgs.tmuxPlugins`. See https://github.com/rycee/home-manager/pull/388/commits/70ca63931dcbe7e4cea6dd0f3550c9032e96b1e0#r219682967

rycee commented:
@stites Sorry for the delay. I pushed some minor changes to your branch. Please give it a try and if it seems ok then I'll merge as-is.

stites commented:
:+1: LGTM!

rycee commented:
Thanks! Squashed and rebased to master in f4ebbcbf707e10b17bf450272dba3bad2c885e6f.
